### DISCUSSION TITLE
Sigh Yet Another Issue with Direct Mode

### DISCUSSION THREAD

**Discussion by Crsarmv7l** [jgromes/RadioLib#1028] [Q&A]
Background: I love my Yardstick One. I have written ~45 different protocol encoders/decoders, nearly every single one of them uses a very raw stream of data with lowball (basically only carrier sense active, PQT at zero, no FEC, whitening or anything else). After looking into it a lot and going through the data sheets, I wanted to create something similar with an Adafruit ESP32-S3 and a CC1101, using synchronous serial to keep things simpler. 

Hardware: 
Adafruit Feather ESP32-S3, [Pinout](https://cdn-learn.adafruit.com/assets/assets/000/110/811/original/adafruit_products_Adafruit_Feather_ESP32-S3_Pinout.png?1649958374)
Ebyte CC1101: [Datasheet with pinout](https://www.ebyte.com/en/downpdf.aspx?id=165)

I also have a Flipper Zero, and the Ebyte CC1101 is fully functional with it hooked up externally

Software:
Arduino IDE 1.8 something
Adafruit board defs are installed
Radiolib installed

Setup:
I am currently only using a single core of the ESP32-S3 for testing (the second will eventually be used to decode rxed signals)

CC1101 - ESP32-S3 Pin Setup:
GDO0 = 6
CSN = 5
SCK = 36
MOSI = 35
MISO/GDO1 = 37
GDO2 = 4

I have also tried:
SCK = 36
MISO/GDO1 = 37
MOSI = 35
CS = 5
GDO0 = 11
GDO2 = 13

Current: Even with a very simple sketch I cannot get any data output aside from [CC1101] Initializing ...Success! So I know it makes it through setup. I should be getting tons of background noise. Where am I going wrong here it is essentially a 1 for 1 of this [sketch](https://github.com/jgromes/RadioLib/blob/cfc425970cbcb2355c0d8c02d6d589228c0d144e/examples/SX127x/SX127x_Receive_Direct/SX127x_Receive_Direct.ino#L4) aside from the few items after radio.begin? Happy to test anything desired.

Below is the sketch:

```
#include <RadioLib.h>

CC1101 radio = new Module(5, 6, RADIOLIB_NC, 4);

const int pin = 4;

void readBit(void) {
  // read the data bit
  radio.readBit(pin);
}

void setup() {
  Serial.begin(19200);
  while (!Serial){};
  Serial.print(F("[CC1101] Initializing ... "));
    
  radio.begin();
  radio.setFrequency(433.92);
  radio.setBitRate(8.00);
  radio.setOOK(true);
  radio.setRxBandwidth(650);
  int state = radio.setPromiscuousMode(true);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Success!"));
  } 
  else {
    Serial.print(F("Failed, Code "));
    Serial.println(state);
    while (true);
  }  

  radio.setDirectAction(readBit);
  radio.receiveDirect();
  
}

void loop() {
  if(radio.available() >= 1) {
    Serial.println("Received packet in direct mode!");
    while(radio.available()) {
      // read a byte
      byte b = radio.read();
  
      // print it
      Serial.print(b, HEX);
      Serial.print('\t');
      Serial.write(b);
      Serial.println();
    }
  }
  
}
```

**Comment by jgromes** [comment]
Couple things:

1. `void readBit(void)` is an interrupt service routine. On ESP32, those have to marked with a special `IRAM_ATTR` attribute for the linker to place them in the correct memeory that will allow their execution.
2. In the example you posted, you are only checking the return value of `setPromiscuousMode`, not the return codes of anything that came before that. So if your CC1101 failed the initialization, you would never know.
3. You did not set any sync word for the direct mode. If you don't want to match with anything, you have to call `setDirectSyncWord(0, 0)`;

**Reply by Crsarmv7l** [reply]
I am still missing something. I implemented state checks all the way and all come back good. I used IRAM_ATTR. Used Promiscuous/didn't use it (doesn't seem like it is really used if sync/async, but I am trying everything). Also tried leaving GDO2 in the module/taking it out no difference. Tried both pin configs (one below is a socketed pcb, the other is a breadboard)

Screenshot:
![image](https://github.com/jgromes/RadioLib/assets/85343771/bf9f0818-a994-492a-a2d5-ec06a610a990)


```
#include <RadioLib.h>

CC1101 radio = new Module(5, 11, RADIOLIB_NC, 13);

const int pin = 13;

void IRAM_ATTR readBit(void) {
  // read the data bit
  radio.readBit(pin);
}

void setup() {
  Serial.begin(19200);
  while (!Serial){};
  Serial.print(F("[CC1101] Initializing ... "));
    
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Success!"));
  } 
  else {
    Serial.print(F("Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  
  state = radio.setFrequency(433.92);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Freq Set!"));
  } 
  else {
    Serial.print(F("Freq Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  
  radio.setBitRate(8.00);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Datarate Set!"));
  } 
  else {
    Serial.print(F("Datarate Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  radio.setOOK(true);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Modulation Set!"));
  } 
  else {
    Serial.print(F("Modulation Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  
  state = radio.setRxBandwidth(650);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Bandwidth Set!"));
  } 
  else {
    Serial.print(F("Bandwidth Failed, Code "));
    Serial.println(state);
    while (true);
  }  

  state = radio.setPromiscuousMode(true);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Promiscuous Set!"));
  } 
  else {
    Serial.print(F("Promiscuous Failed, Code "));
    Serial.println(state);
    while (true);
  } 

  state = radio.setDirectSyncWord(0, 0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Direct SyncWord Set!"));
  } 
  else {
    Serial.print(F("Direct SyncWord Failed, Code "));
    Serial.println(state);
    while (true);
  }
  
  radio.setDirectAction(readBit);
  radio.receiveDirect();
  
}

void loop() {
  if(radio.available() >= 1) {
    Serial.println("Received packet in direct mode!");
    while(radio.available()) {
      // read a byte
      byte b = radio.read();
  
      // print it
      Serial.print(b, HEX);
      Serial.print('\t');
      Serial.write(b);
      Serial.println();
    }
  }
  
}
```

**Comment by Crsarmv7l** [comment]
Had some more thinks:
I needed to set PQT =  0. I am getting data now. Haven't verified it is actual data, but progress!

A side note. It is very strange to have to set a preamble length (min of 16 allowed value) in order to set PQT, and then remove the preamble. It would be really nice to be able to set PQT in setPromiscuousMode since that is disabling preamble, syncword, etc (would also be really nice and make sense to also be able to set carriersense in Promiscuous too, but not as important)

```
#include <RadioLib.h>

CC1101 radio = new Module(5, 11, RADIOLIB_NC, 13);

const int pin = 13;

void IRAM_ATTR readBit(void) {
  // read the data bit
  radio.readBit(pin);
}

void setup() {
  Serial.begin(19200);
  while (!Serial){};
  Serial.print(F("[CC1101] Initializing ... "));
    
  int state = radio.begin();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Success!"));
  } 
  else {
    Serial.print(F("Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  
  state = radio.setFrequency(433.92);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Freq Set!"));
  } 
  else {
    Serial.print(F("Freq Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  
  radio.setBitRate(8.00);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Datarate Set!"));
  } 
  else {
    Serial.print(F("Datarate Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  radio.setOOK(true);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Modulation Set!"));
  } 
  else {
    Serial.print(F("Modulation Failed, Code "));
    Serial.println(state);
    while (true);
  }  
  
  state = radio.setRxBandwidth(650);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Bandwidth Set!"));
  } 
  else {
    Serial.print(F("Bandwidth Failed, Code "));
    Serial.println(state);
    while (true);
  }  

  state = radio.setPreambleLength(16, 0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("PQT Set!"));
  } 
  else {
    Serial.print(F("PQT Failed, Code "));
    Serial.println(state);
    while (true);
  } 

  state = radio.setPromiscuousMode(true);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Promiscuous Set!"));
  } 
  else {
    Serial.print(F("Promiscuous Failed, Code "));
    Serial.println(state);
    while (true);
  } 

  state = radio.setDirectSyncWord(0, 0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Direct SyncWord Set!"));
  } 
  else {
    Serial.print(F("Direct SyncWord Failed, Code "));
    Serial.println(state);
    while (true);
  }
  
  radio.setDirectAction(readBit);
  radio.receiveDirect();
  
}

void loop() {
  if(radio.available() >= 1) {
    Serial.println("Received packet in direct mode!");
    while(radio.available()) {
      // read a byte
      byte b = radio.read();
  
      // print it
      Serial.print(b, HEX);
      Serial.print('\t');
      Serial.write(b);
      Serial.println();
    }
  }
  
}
'''

**Reply by jgromes** [reply]
Glad you were able to make some progress! Regarding PQT, I think we can modify `setPromiscuousMode` so that PQT is set to 0 when `setPromiscuousMode(true)` is called, is that what you mean?

**Reply by Crsarmv7l** [reply]
Yeah, it makes sense to do so, otherwise promiscuous won't return anything like I saw. I forked radiolib after posting that and did it. Haven't tested yet but it should work:

![image](https://github.com/jgromes/RadioLib/assets/85343771/ff61397b-add5-4c9c-b986-8d759463b4f0)

I also later enabled carriersense in Promiscuous with DisableSyncWordFiltering(true). Fine for me, but my fix wouldn't be good for here as there are two options that make sense with  the Promiscuous; RADIOLIB_CC1101_SYNC_MODE_NONE_THR  and RADIOLIB_CC1101_SYNC_MODE_NONE. None of the other Sync Modes are valid with Promiscuous.

Would be nice to add another arg to Promiscuous for those two. Eg radio.setPromiscuous(Bool enable, Bool Carriersense) would cover both Sync Modes.

Edit: Did the carrier sense bool for promiscuous as well, instead of just DisableSyncWordFiltering(true). If you are open to both changes I will cancel the PQT = 0 only pull request and submit both.
